### ISSUE TITLE
Quantum.add_job("name", job) API does not work

### ISSUE THREAD

**Issue by dapdizzy** [c-rack/quantum-elixir#89]
Dudes, hu there again!

Weird, but Quantum.add_job("name", job) where job is %Quantum.Job{} does not work.

First off, everithing works fine when I add a job like this:
Quantum.add_job("* * * * *", fn -> IO.puts "Tick" end)

so I have :quantum app up and running and nothins is wrong with it.

Now, when I use the add_job API passing it the %Quantum.Job{} struct as you have carefully described in the readme.md, same stuff does not work.

Any clues?

I mean it looks really weird as this seems to be the very basic stuff that should just work if the quantum app is up and running.

I'm using [**{:quantum, ">= 1.7.0"}]** dependency.

Any pointers are highly appreciated. I'm also using the timezone: :local option if that is important.

Thanks, Dmitry.

**Comment by c-rack** [comment]
Could you please provide a minimal test case that reproduces the problem, @dapdizzy ?

Maybe the tests for this function are a good starting point:
https://github.com/c-rack/quantum-elixir/blob/master/test/quantum_test.exs#L18-L28

**Comment by dapdizzy** [comment]
I'm exposing some bad practices and don't write test for now, anyway, the thing is, when I write this:
```
m = __MODULE__
f = :ping
a = [someargs]
job =
    %Quantum.Job
    {
      schedule: schedule,
      task: {m, f},
      args: a,
      name: "pinger",
      nodes: [node()]
    }
Quantum.add_job(job.name, job)
```
nothing is fired at the expected schedule.
When I change it to the following:
`Quantum.add_job(schedule, fn -> apply(m, f, a) end)`
everything works as expected.
A do pass a tumlpe {m, f} to as a task, as opposed to passing a function reference, as I can see this kind of argument is used in the example in [this section](https://github.com/c-rack/quantum-elixir#named-jobs).

Another thing I want to mention is a need to extent the readme to reaaly provide you with some help on the Quantum API at least. For instance, from yourexample of a test case I learned that Quantum has a function called jobs that return an enumerable of jobs which is somewhat important to me and it is not convenient enough that is is not briefly described in the readme.

If it could really help, I'll probably be able to provide a working test case, but fow not I kind of sticked to the working API as a workaround.

**Comment by c-rack** [comment]
Regarding your comment about the missing `Quantum.jobs` reference in the README:

This function is documented in our API docs here:
https://hexdocs.pm/quantum/Quantum.html#jobs/0

Not every API function needs to be mentioned in the README.
That is what an API documentation is for ;-)

And yes, providing a failing test case would really help to investigate your issue, thanks!

**Comment by dapdizzy** [comment]
I love the way you put things in order in my head. I wasn't aware that hep is the place to look for the docs.Putting the link to the docs in the readme would make it a smooth experience for the user.

I'll take my time to make a test case prooving or disproving my issue.

Thanks for your response about the API docs!

**Comment by dapdizzy** [comment]
Taking a short look at the test cases on the page that you've provided a link to, I cannot see the usage of a tuple `{Module, :function}` as a value for the `%Quantum.Job{task: task}` Just wondering: is there a test case in your suite for this king of pattern? Is it a valid use to pass a `{Module, :funcvtion}` as a value for the `task` item in the `%Quantum.Job{}` struct? I'm refering to the readme (the link that I've sent you)where rthis kind of usage seems to be promoted.

Anyways, I'll try to provide a testcase as a tangible basis prooving the issue.

**Comment by dapdizzy** [comment]
Also, test I can see seem to assert job inclusion in the jobs list. What I want to test is job execution. I know it's more king of a feature-level test, rather then an atomic functional test, but that's what I'm really concerned about. Any clues on how shoud I write such test? Mor specifically, how the assert statement should look like?

**Comment by dapdizzy** [comment]
Run a kind of complete test scenario and seems like everything works fine.I havn't figured out why it didn;t irk in the scenario I was working on in that partucular context, but I guess the issue can be closed as 'not reproduced' or something.

Just looking at the code that did not have any effect I did exactly the same thing, although I was using __MODULE__ for the module instead of providint a literal name. Other than that I see no differences.

Anyways, thatks for your responsiveness. I think even such false alerts contribute to making the project more 'alive' and prone to more production use cases.

Thanks Constantin! Best regards, Dmitry.

**Comment by dapdizzy** [comment]
I think there's a bug in the way default value for the nodes is being embedded into the module attribute in Quantum.Job module.

It is defaulted to the module attribute value `@defaultnodes` which seems to be evaluated too early and always defaults to nonode@nohost in my case, whereas the app is started in a named node (using --sname), thus nothing is fired at the scheduled time. When I start to explicitly define nodes in the Job struct, it starts to work as expected.

Could you please look into the issue?

**Comment by c-rack** [comment]
@dapdizzy this is a new issue and unrelated to the original one, so I close this again.

Please open a new issue according to our [contribution process](https://github.com/c-rack/quantum-elixir#contribution), thanks.

**Comment by dapdizzy** [comment]
Constantin, just a note: although the issue as described does not seem to
be related with the original one, when using the API that takes a job
strict directly, it looks like it is not working when ran on a named node.
Anyway, I've opened a new issue with pretty technical description on the
problem and easy reproduce steps making the issue show up.

Best regards, Dmitry.

вторник, 22 марта 2016 г. пользователь Constantin Rack написал:

**Comment by c-rack** [comment]
Thanks, Dmitry!

**Comment by dapdizzy** [comment]
Constantin, I've sent a pull request fixing nodes defaulting. It works for
me. Would be nice, if you'd take a look and we might get the issue closed.

Thanks!

Best regards, Dmitry.

2016-03-22 16:36 GMT+03:00 Constantin Rack <notifications@github.com>:

**Comment by c-rack** [comment]
Thanks. Just 2 minor things:

1. Your PR breaks the Travis CI build, would you mind to take a look and fix them?
2. Could you please include a "Fixes #89" into your PR description or one of the commits? That way, this issue will be automatically closed by merge:
https://help.github.com/articles/closing-issues-via-commit-messages/

**Comment by dapdizzy** [comment]
Yeah saw it recently)
I'll fix it during the day and I'll look into the second point that you've
mentioned.

Best regards, Dmitry.

среда, 23 марта 2016 г. пользователь Constantin Rack написал:
